apps/prairielearn/src/lib/externalGraderResults.js

```diff
@@ -169,7 +169,7 @@ async function processMessage(data) {
         ResponseContentType: 'application/json',
       });
       if (!data.Body) throw new Error('No body in S3 response');
-      await processResults(jobId, data.Body.transformToString());
+      await processResults(jobId, await data.Body.transformToString());
```

Comment:
For lack of a better place to comment this: I'm wondering if we should better handle this kind of internal error via `externalGraderCommon.makeGradingResult()`. In this case, the promise became `{}` after `JSON.serialize`, which explains why this didn't error outright. However, I think we should validate the data and if, for instance, there isn't a top-level `succeeded` field, we should error more explicitly.

I'm also wondering if these kinds of internal errors should mark the submission as ungradable so that students aren't penalized for points/attempts.

Reply:
I agree with both of these ideas. Running the results through `makeGradingResult()` seems very helpful for more precisely identifying the source of errors and it's true that `ungradable` seems like the correct result when we didn't clearly get a score of 0.

Reply:
👍 I'll make these changes in a quick followup PR.